File: menuItem.cpp
#include "menuItem.hpp"
#include <iostream>
#include <iomanip>

MenuItem::MenuItem()
	: m_name("")
	, m_costPrice(0.0)
	, m_sellingPrice(0.0)
	, m_techCard()
	, m_salesCount(0)
	, m_isAvailable(true)
{
}

MenuItem::MenuItem(const std::string& name, double costPrice, double sellingPrice, bool isAvailable)
	: m_name(name)
	, m_costPrice(costPrice)
	, m_sellingPrice(sellingPrice)
	, m_techCard()
	, m_salesCount(0)
	, m_isAvailable(isAvailable)
{
}

// Конструктор копирования
MenuItem::MenuItem(const MenuItem& other)
	: m_name(other.m_name)
	, m_costPrice(other.m_costPrice)
	, m_sellingPrice(other.m_sellingPrice)
	, m_techCard(other.m_techCard)
	, m_salesCount(other.m_salesCount)
	, m_isAvailable(other.m_isAvailable)
{
}

double MenuItem::CalculateProfitPerUnit() const
{
	return m_sellingPrice - m_costPrice;
}

double MenuItem::CalculateTotalProfit() const
{
	return CalculateProfitPerUnit() * m_salesCount;
}

double MenuItem::CalculateMargin() const
{
	if (m_sellingPrice == 0.0)
	{
		return 0.0;
	}
	return (CalculateProfitPerUnit() / m_sellingPrice) * 100.0;
}

void MenuItem::PrintInfo() const
{
	std::cout << "=== Информация о блюде ===" << std::endl;
	std::cout << "Наименование: " << m_name << std::endl;
	std::cout << "Себестоимость: " << std::fixed << std::setprecision(2) << m_costPrice << std::endl;
	std::cout << "Цена продажи: " << m_sellingPrice << std::endl;
	std::cout << "Количество продаж: " << m_salesCount << std::endl;
	std::cout << "Доступно к продаже: " << (m_isAvailable ? "Да" : "Нет") << std::endl;
	std::cout << "Прибыль с единицы: " << CalculateProfitPerUnit() << std::endl;
	std::cout << "Общая прибыль: " << CalculateTotalProfit() << std::endl;
	std::cout << "Маржинальность: " << CalculateMargin() << "%" << std::endl;
	std::cout << "=========================" << std::endl;
}

void MenuItem::SetAvailability(bool isAvailable)
{
	m_isAvailable = isAvailable;
}

std::string MenuItem::GetName() const
{
	return m_name;
}

double MenuItem::GetCostPrice() const
{
	return m_costPrice;
}

double MenuItem::GetSellingPrice() const
{
	return m_sellingPrice;
}

MenuItem::TechCardType MenuItem::GetTechCard() const
{
	return m_techCard;
}

int MenuItem::GetSalesCount() const
{
	return m_salesCount;
}

bool MenuItem::IsAvailable() const
{
	return m_isAvailable;
}

void MenuItem::SetName(const std::string& name)
{
	m_name = name;
}

void MenuItem::SetCostPrice(double costPrice)
{
	m_costPrice = costPrice;
}

void MenuItem::SetSellingPrice(double sellingPrice)
{
	m_sellingPrice = sellingPrice;
}

void MenuItem::SetTechCard(const TechCardType& techCard)
{
	m_techCard = techCard;
}

void MenuItem::SetSalesCount(int salesCount)
{
	m_salesCount = salesCount;
}

// Реализация перегруженных операторов для MenuItem
MenuItem& MenuItem::operator+=(int salesCount)
{
	if (salesCount > 0)
	{
		this->m_salesCount += salesCount;  // Использование this для доступа к полю
	}
	return *this;  // Возврат ссылки на текущий объект
}

bool MenuItem::operator<(const MenuItem& other) const
{
	return this->m_sellingPrice < other.m_sellingPrice;  // Использование this для сравнения
}

bool MenuItem::operator>(const MenuItem& other) const
{
	return this->m_sellingPrice > other.m_sellingPrice;  // Использование this для сравнения
}

// Реализация дружественной функции: расчет общей стоимости меню
double CalculateTotalMenuCost(const std::vector<MenuItem>& menuItems)
{
	double totalCost = 0.0;
	// Дружественная функция имеет доступ к приватным полям
	for (const auto& item : menuItems)
	{
		if (item.m_isAvailable)  // Учитываем только доступные блюда
		{
			totalCost += item.m_sellingPrice;
		}
	}
	return totalCost;
}

File: lab4/MenuItem.java
import java.util.ArrayList;
import java.util.List;

public class MenuItem {
    // Тип для технологической карты (список списков строк)
    // Каждый внутренний список представляет строку технологической карты
    public static class TechCardType extends ArrayList<List<String>> {
        // Наследуемся от ArrayList для удобства работы
    }

    // Приватные поля
    private String name;                  // Название блюда
    private double costPrice;             // Себестоимость
    private double sellingPrice;         // Цена продажи
    private TechCardType techCard;       // Технологическая карта
    private int salesCount;               // Количество продаж
    private boolean isAvailable;         // Доступно к продаже

    // Конструктор по умолчанию
    public MenuItem() {
        this.name = "";
        this.costPrice = 0.0;
        this.sellingPrice = 0.0;
        this.techCard = new TechCardType();
        this.salesCount = 0;
        this.isAvailable = true;
    }

    // Конструктор с параметрами
    public MenuItem(String name, double costPrice, double sellingPrice, boolean isAvailable) {
        this.name = name;
        this.costPrice = costPrice;
        this.sellingPrice = sellingPrice;
        this.techCard = new TechCardType();
        this.salesCount = 0;
        this.isAvailable = isAvailable;
    }

    // Расчет прибыли с единицы
    public double calculateProfitPerUnit() {
        return sellingPrice - costPrice;
    }

    // Расчет общей прибыли
    public double calculateTotalProfit() {
        return calculateProfitPerUnit() * salesCount;
    }

    // Расчет маржинальности (в процентах)
    public double calculateMargin() {
        if (sellingPrice == 0.0) {
            return 0.0;
        }
        return (calculateProfitPerUnit() / sellingPrice) * 100.0;
    }

    // Вывести информацию о блюде
    public void printInfo() {
        System.out.println("=== Информация о блюде ===");
        System.out.println("Наименование: " + name);
        System.out.printf("Себестоимость: %.2f%n", costPrice);
        System.out.printf("Цена продажи: %.2f%n", sellingPrice);
        System.out.println("Количество продаж: " + salesCount);
        System.out.println("Доступно к продаже: " + (isAvailable ? "Да" : "Нет"));
        System.out.printf("Прибыль с единицы: %.2f%n", calculateProfitPerUnit());
        System.out.printf("Общая прибыль: %.2f%n", calculateTotalProfit());
        System.out.printf("Маржинальность: %.2f%%%n", calculateMargin());
        System.out.println("=========================");
    }

    // Установить доступность
    public void setAvailability(boolean isAvailable) {
        this.isAvailable = isAvailable;
    }

    // Геттеры
    public String getName() {
        return name;
    }

    public double getCostPrice() {
        return costPrice;
    }

    public double getSellingPrice() {
        return sellingPrice;
    }

    public TechCardType getTechCard() {
        return techCard;
    }

    public int getSalesCount() {
        return salesCount;
    }

    public boolean isAvailable() {
        return isAvailable;
    }

    // Сеттеры
    public void setName(String name) {
        this.name = name;
    }

    public void setCostPrice(double costPrice) {
        this.costPrice = costPrice;
    }

    public void setSellingPrice(double sellingPrice) {
        this.sellingPrice = sellingPrice;
    }

    public void setTechCard(TechCardType techCard) {
        this.techCard = techCard;
    }

    public void setSalesCount(int salesCount) {
        this.salesCount = salesCount;
    }

    // Метод для увеличения счетчика продаж (будет использоваться при завершении заказа)
    public void incrementSalesCount() {
        // TODO: будет связано с Order.CompleteOrder() и Employee.addSales()
        this.salesCount++;
    }
}

